lib/datadog/di/probe_manager.rb

```diff
@@ -0,0 +1,182 @@
+      attr_reader :installed_probes
+      attr_reader :pending_probes
+
+      # Probes that failed to instrument for reasons other than the target is
+      # not yet loaded are added to this collection, so that we do not try
+      # to instrument them every time remote configuration is processed.
+      attr_reader :failed_probes
+
+      # config is one probe info
+      def add_probe(probe)
+        # TODO lock here?
+
+        # Probe failed to install previously, do not try to install it again.
+        if msg = failed_probes[probe.id]
+          # TODO test this path
+          raise Error::ProbePreviouslyFailed, msg
+        end
+
+        begin
+          instrumenter.hook(probe, &method(:probe_executed_callback))
+
+          installed_probes[probe.id] = probe
+          payload = probe_notification_builder.build_installed(probe)
+          probe_notifier_worker.add_status(payload)
+          # The probe would only be in the pending probes list if it was
+          # previously attempted to be installed and the target was not loaded.
+          # Always remove from pending list here because it makes the
+          # API smaller and shouldn't cause any actual problems.
+          pending_probes.delete(probe.id)
+          true
+        rescue Error::DITargetNotDefined => exc
+          pending_probes[probe.id] = probe
+          false
+        end
+      rescue => exc
+        # In "propagate all exceptions" mode we will try to instrument again.
+        raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+
+        logger.warn("Error processing probe configuration: #{exc.class}: #{exc}")
+        telemetry&.report(exc, description: "Error processing probe configuration")
+        # TODO report probe as failed to agent since we won't attempt to
+        # install it again.
+
+        # TODO add top stack frame to message
+        failed_probes[probe.id] = "#{exc.class}: #{exc}"
+
+        raise
+      end
+
+      def remove_other_probes(probe_ids)
+        pending_probes.values.each do |probe|
+          unless probe_ids.include?(probe.id)
+            pending_probes.delete(probe.id)
+          end
+        end
+        installed_probes.values.each do |probe|
+          unless probe_ids.include?(probe.id)
+            begin
+              instrumenter.unhook(probe)
+              # Only remove the probe from installed list if it was
+              # successfully de-instrumented. Active probes do incur overhead
+              # for the running application, and if the error is ephemeral
+              # we want to try removing the probe again at the next opportunity.
+              #
+              # TODO give up after some time?
+              installed_probes.delete(probe.id)
+            rescue => exc
+              raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+              # Silence all exceptions?
+              # TODO should we propagate here and rescue upstream?
```

Comment:
Trying to answer this comment: in my read of this PR, it makes sense to try to `unhook` all probes, even if one of them fail to unhook. Which would mean that propagating the error early will prevent other probes from being propagated. So it looks to me that we should not propagate the error immediately here.